src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrProviderStateFactory.java

```diff
@@ -167,18 +172,28 @@ private JcrProviderState createJcrProviderState(
      * @param logoutSession
      *            whether to logout the <code>session</code> after impersonation
      *            or not.
+     * @param explicitSessionUsed
+     *            whether the JCR session was explicitly given in the auth info or not.
      * @return The original session or impersonated session.
      * @throws LoginException
      *             If something goes wrong.
      */
     private static Session handleImpersonation(final Session session, final Map<String, Object> authenticationInfo,
-            final boolean logoutSession) throws LoginException {
+            final boolean logoutSession, boolean explicitSessionUsed) throws LoginException {
         final String sudoUser = getSudoUser(authenticationInfo);
-        if (sudoUser != null && !session.getUserID().equals(sudoUser)) {
+        // Do we need session.impersonate() because we are asked to impersonate another user?
+        boolean needsSudo = (sudoUser != null) && !session.getUserID().equals(sudoUser);
+        // Do we need session.impersonate() to get an independent copy of the session we were given in the auth info?
+        boolean needsCloning = !needsSudo && explicitSessionUsed && authenticationInfo.containsKey(ResourceProvider.AUTH_CLONE);
+        if (needsCloning || needsSudo) {
+            // If we just need to clone the session, we impersonate with the same user ID and not set an impersonator attribute.
+            // In all other cases, it's a "proper" sudo to the given user.
             try {
-                final SimpleCredentials creds = new SimpleCredentials(sudoUser, new char[0]);
+                final SimpleCredentials creds = new SimpleCredentials(needsSudo ? sudoUser : session.getUserID(), new char[0]);
                 copyAttributes(creds, authenticationInfo);
-                creds.setAttribute(ResourceResolver.USER_IMPERSONATOR, session.getUserID());
+                if (needsSudo) {
+                    creds.setAttribute(ResourceResolver.USER_IMPERSONATOR, session.getUserID());
+                }
```

Comment:
I would put the `needsSudo` handling together in one place, so it's more obvious what's going on:

```
SimpleCredentials creds;
if (needsSudo) {
    creds = new SimpleCredentials(sudoUser, new char[0]);
    copyAttributes(creds, authenticationInfo);
    creds.setAttribute(ResourceResolver.USER_IMPERSONATOR, session.getUserID());
} else {
    creds = new SimpleCredentials(session.getUserID(), new char[0]);
    copyAttributes(creds, authenticationInfo);
}
return session.impersonate(creds);
```

(I am assuming the `setAttribute()` for the `USER_IMPERSONATOR` must happen after the `copyAttributes()` to ensure it overwrites anything that might be in the authenticationInfo.)

Reply:
Hmm, at this point, we would have a nested if/try/if just to avoid repeating one common line, so maybe it's more readable to share nothing:

```java
try {
    if (needsSudo) {
        SimpleCredentials creds = new SimpleCredentials(sudoUser, new char[0]);
        copyAttributes(creds, authenticationInfo);
        creds.setAttribute(ResourceResolver.USER_IMPERSONATOR, session.getUserID());
        return session.impersonate(creds);
    } else if (needsCloning) {
        SimpleCredentials creds = new SimpleCredentials(session.getUserID(), new char[0]);
        copyAttributes(creds, authenticationInfo);
        return session.impersonate(creds);
    }
} catch (final RepositoryException re) {
...
```
This way we don't need the outermost `if (needsSudo || needsCloning)` so there is only two levels of nesting.

Reply:
👍 even better